### DISCUSSION TITLE
browser-based client-side tx building

### DISCUSSION THREAD

**Discussion by waalge** [input-output-hk/cardano-js-sdk#1059] [General]
I'm looking at using the cardano-js-sdk as a framework for browser-based client-side tx building. 
This will include plutus scripts and the additional capabilities that entails. 

What are your thoughts on this?

Aiken can provide the plutus-script evaluator and fee calculator. I've played with it a bit to see what else might be missing. 
I think the interface to the `TxBuilder` might need to be extended to include specifying inputs, adding validity ranges. However, I might be missing how it works (eg the way input selection can be setup to ensure the spend of certain inputs).

**Comment by rhyslbw** [comment]
We have this in mind too and would welcome contributions

**Comment by waalge** [comment]
Not a great ts dev here(*). Would probably be best for the core team to map out and propose an api. 

Current apis: lucid, pycardano, helios, _etc_ 

I currently use lucid, and think it leaves much room for improvement (a more direct correspondence to the CML, for example). Cant comment on the usability of the other two listed. Tx building can be quite bespoke, 
so an api with full control available for the 5% of cases where its needed, and at the same time convenient for the 95% of other cases. 


*Struggled to get the thing running. https://github.com/input-output-hk/cardano-js-sdk/issues/1058

**Comment by rhyslbw** [comment]
Have you seen https://meshjs.dev?

> Tx building can be quite bespoke

Yes we agree with that!

**Reply by waalge** [reply]
Of course. Although some time back. Its hard to keep up. 

For meshjs, the last time i interacted with devs, we were discussing CIP57/blueprint integration. I don't know if this is present/ ergonomic. I also don't think the builder can estimate exunits/ fees. Two things lucid has that massively improve DX.

I'm watching `blaze`. I think they got quite a few proposals through catalyst. Hopefully that resolves a lot of the "not yet implemented"

On cardano-js-sdk, the tx construction lib has snake case directories. One of those high level things which doesn't inspire confidence. I saw angelo's inclusion script support for spends. I'll keep watching.   

After a very meandering route trying many things, I've ended up coping with patching lucid.
